ts2mls/js/src/main/scala/ts2mls/TSCompilerInfo.scala

```diff
@@ -0,0 +1,174 @@
+  // e.g. `function f(x: int = 42) {}`, we can use it directly: `f()`.
+  // in this case, the initializer would store the default value and be non-empty.
+  lazy val isOptionalParameter = checker.isOptionalParameter(node)
+  lazy val isStatic = if (modifiers.isUndefined) false
+                     else modifiers.foldLeft(false)((s, t) => t.isStatic)
+
+  lazy val typeNode = TSTypeObject(checker.getTypeFromTypeNode(node))
+  lazy val symbol = TSSymbolObject(node.symbol)
+  lazy val parameters = TSNodeArray(node.parameters)
+  lazy val typeParameters = TSNodeArray(node.typeParameters)
+  lazy val constraint = TSTokenObject(node.constraint)
+  lazy val members = TSNodeArray(node.members)
+  lazy val types = TSNodeArray(node.types)
+  lazy val heritageClauses = TSNodeArray(node.heritageClauses)
+  lazy val initializer = TSNodeObject(node.initializer)
+  lazy val modifier =
+    if (modifiers.isUndefined) Public
+    else modifiers.foldLeft[TSAccessModifier](Public)(
+      (m, t) => if (t.isPrivate) Private else if (t.isProtected) Protected else m)
+
+  lazy val locals = TSSymbolMap(node.locals)
+  lazy val returnType = TSTypeObject(checker.getReturnTypeOfSignature(node))
+  lazy val `type` = TSNodeObject(node.selectDynamic("type"))
+
+  lazy val symbolType = TSTypeObject(checker.getTypeOfSymbolAtLocation(node.symbol, node))
+}
+
+object TSNodeObject {
+  def apply(node: js.Dynamic)(implicit checker: TSTypeChecker) = new TSNodeObject(node)
+}
+
+class TSTokenObject(token: js.Dynamic)(implicit checker: TSTypeChecker) extends TSAny(token) {
+  private lazy val kind = token.kind
+
+  lazy val isPrivate = kind == TypeScript.syntaxKindPrivate
+  lazy val isProtected = kind == TypeScript.syntaxKindProtected
+  lazy val isStatic = kind == TypeScript.syntaxKindStatic
+
+  lazy val typeNode = TSTypeObject(checker.getTypeFromTypeNode(token))
+}
+
+object TSTokenObject {
+  def apply(token: js.Dynamic)(implicit checker: TSTypeChecker) = new TSTokenObject(token)
+}
+
+class TSTypeObject(obj: js.Dynamic)(implicit checker: TSTypeChecker) extends TSAny(obj) {
+  private lazy val flags = obj.flags
+  private lazy val objectFlags = if (IsUndefined(obj.objectFlags)) 0 else obj.objectFlags
+
+  lazy val symbol = TSSymbolObject(obj.symbol)
+  lazy val typeArguments = TSTypeArray(checker.getTypeArguments(obj))
+  lazy val intrinsicName = obj.intrinsicName.toString
+  lazy val types = TSTypeArray(obj.types)
+  lazy val properties = TSSymbolArray(checker.getPropertiesOfType(obj))
+  lazy val node = TSNodeObject(checker.typeToTypeNode(obj))
+  lazy val elementTypeOfArray = TSTypeObject(checker.getElementTypeOfArrayType(obj))
+
+  lazy val isTupleType = node.isTupleTypeNode
+  lazy val isArrayType = node.isArrayTypeNode
+  lazy val isEnumType = (flags & TypeScript.typeFlagsEnumLike) > 0
+  lazy val isUnionType = obj.isUnion().asInstanceOf[Boolean]
+  lazy val isIntersectionType = obj.isIntersection().asInstanceOf[Boolean]
```

Comment:
All unsafe cases (`asInstanceOf`) should be removed before the merge. Find a better/cleaner way.

Reply:
All `asInstanceOf` have been removed.